### PR TITLE
implement rds api

### PR DESCRIPTION
**Pull request by mattklein123**
fixes https://github.com/lyft/envoy/issues/298
fixes https://github.com/lyft/envoy/issues/463

### CODE REVIEW

**Comment by mattklein123** [comment]
@lyft/network-team @rshriram @GregHanson

File: docs/configuration/http_conn_man/rds.rst [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+cluster
+  *(required, string)* The name of an upstream :ref:`cluster <config_cluster_manager_cluster>` that
+  hosts the route discovery service. The cluster must run a REST service that implements the
+  :ref:`RDS HTTP API <config_http_conn_man_rds_api>`.
+
+route_config_name
+  *(required, string)* The name of the route table. This name will be passed to the
+  :ref:`RDS HTTP API <config_http_conn_man_rds_api>` so that different route tables can be used by
```
Comment (ccaraman) [inline docs/configuration/http_conn_man/rds.rst]: What does it mean to be used by independent filters?

File: include/envoy/router/rds.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * A provider for constant route configurations.
+ */
+class RouteConfigProvider {
+public:
+  virtual ~RouteConfigProvider() {}
+
+  /**
+   * @return Router::ConfigPtr a route configuration for use during a single request. The returned
+   *         config may be different on a subsequent call so a new config should be acquired for
```
Comment (ccaraman) [inline include/envoy/router/rds.h]: comma before so

File: source/common/json/config_schemas.cc [hunk 3]
```diff
@@ -242,7 +260,7 @@ const std::string Json::Schema::HTTP_CONN_NETWORK_FILTER_SCHEMA(R"EOF(
       "use_remote_address" : {"type" : "boolean"},
       "generate_request_id" : {"type" : "boolean"}
     },
-    "required" : ["codec_type", "stat_prefix", "route_config", "filters"],
+    "required" : ["codec_type", "stat_prefix", "filters"],
```
Comment (ccaraman) [inline source/common/json/config_schemas.cc]: I think you should keep route_config. At line 222, it only checks for an object, so empty will pass. And the documentation above makes it seem like it should always be there.

Comment (mattklein123) [inline source/common/json/config_schemas.cc]: It's not required anymore. I'm fixing the docs.

File: source/common/router/rds_impl.cc [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  request.headers().insertPath().value(fmt::format("/v1/routes/{}/{}/{}", route_config_name_,
+                                                   local_info_.clusterName(),
+                                                   local_info_.nodeName()));
+}
+
+void RdsRouteConfigProviderImpl::parseResponse(const Http::Message& response) {
+  log_debug("rds: parsing response");
+  Json::ObjectPtr response_json = Json::Factory::LoadFromString(response.bodyAsString());
+  uint64_t new_hash = response_json->hash();
+  if (new_hash != last_config_hash_ || !initialized_) {
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: I will add some detail in the docs. It's just a performance optimization for the route config to now reload internally, so I don't think it's worth guiding developers to not reorder things.

**Comment by mattklein123** [comment]
@ccaraman @rshriram updated from comments. Also, I opened https://github.com/lyft/envoy/issues/468 to do a full overview of all of the configuration options now available. I will do this as an independent change.

File: source/common/router/rds_impl.cc (hunk 4 above)
Comment (rshriram) [inline source/common/router/rds_impl.cc]: Sorry, yeah you are right. It was more for istio, where we are making sure that the config remains as deterministic as possible between every refresh.

File: docs/configuration/network_filters/client_ssl_auth_filter.rst [hunk 5]
```diff
@@ -13,7 +13,7 @@ Client TLS authentication filter :ref:`architecture overview <arch_overview_ssl_
     "config": {
       "auth_api_cluster": "...",
       "stat_prefix": "...",
-      "refresh_interval_ms": "...",
+      "refresh_delay_ms": "...",
```
Comment (ccaraman) [inline docs/configuration/network_filters/client_ssl_auth_filter.rst]: I just realized that CLIENT_SSL_NETWORK_FILTER_SCHEMA doesn't have `refresh_delay_ms`. Can you please add that?

Comment (mattklein123) [inline docs/configuration/network_filters/client_ssl_auth_filter.rst]: fixed

File: docs/configuration/http_conn_man/rds.rst [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  value is 30000ms (30 seconds).
+
+.. _config_http_conn_man_rds_api:
+
+REST API
+--------
+
+.. http:get:: /v1/routes/(string: route_config_name)/(string: service_cluster)/(string: service_node)
+
+Asks the discovery service to return the route configuration for a particular `route_config_name`,
```
Comment (junr03) [inline docs/configuration/http_conn_man/rds.rst]: Asks the *route* dicovery service... People might get confused.

File: test/common/router/config_impl_test.cc [hunk 7]
```diff
@@ -687,7 +699,59 @@ TEST(RouteMatcherTest, ShadowClusterNotFound) {
...
+  ]
+}
+  )EOF";
+
+  Json::ObjectPtr loader = Json::Factory::LoadFromString(json);
+  NiceMock<Runtime::MockLoader> runtime;
+  NiceMock<Upstream::MockClusterManager> cm;
+  EXPECT_CALL(cm, get("www2")).WillRepeatedly(Return(nullptr));
+
+  ConfigImpl(*loader, runtime, cm, false);
```
Comment (ccaraman) [inline test/common/router/config_impl_test.cc]: nit: add a EXPECT_NO_THROW

**Pull request merged**